File: C++/22.括号生成.cpp
/*
 * @lc app=leetcode.cn id=22 lang=cpp
 *
 * [22] 括号生成
 */

// @lc code=start
class Solution {
public:
    vector<string> res;
    vector<string> generateParenthesis(int n) {
        dfs("", 0, 0, n);
        return res;
    }
    void dfs(string istr, int left, int right, int n) {
        if(left == n && right == n){
            res.push_back(istr);
            return;
        }
            
        if(left < n)
            dfs(istr+"(", left+1, right, n);
        if(left > right && right < n)
            dfs(istr+")", left, right+1, n);
    }
};

// @lc code=end

File: Go/22.括号生成.go
/*
 * @lc app=leetcode.cn id=22 lang=golang
 *
 * [22] 括号生成
 */

// @lc code=start
func generateParenthesis(n int) []string {
	res := []string{}
	sub := ""
	backtrack(n, sub, &res, 0, 0)
	return res
}

func backtrack(n int, sub string, res *[]string, left, right int) {
	if right == n && left == n {
		*res = append(*res, sub)
		return
	}

	if left < n {
		backtrack(n, sub+"(", res, left+1, right)
	}
	if left > right && right < n {
		backtrack(n, sub+")", res, left, right+1)
	}
}

// @lc code=end

